### PR TITLE
[MRG+1] Add support for meaningful exog names 

### LINKED ISSUE #222: meaningful exog_names for the fourier terms variables generated by FourierFeaturizer
when asking for a `summary()` the fourier terms generated by `FourierFeaturizer` should have meaningful names like in R with their period and their order, (e.g. `S24-5` and `C24-5`) and not `x14` like now. 

this shouldn't be so hard, as for my own exogenous variables I'm already doing it, after the fit with `model.arima_res_.model.exog_names[1:1+len(ex_cols)] = ex_cols`

### PR DESCRIPTION
**Pull request by tgsmith61591**
# Description

Fixes #222. Adds an optional parameter, `prefix`, to exogenous featurizers and pass-through support for pandas DataFrames as exog arrays (rather than converting to np.ndarray). This accomplishes two things:

* If a pandas frame is provided as an exog array to the statsmodels SARIMAX class, it will keep the names intact and the `summary` method will contain the meaningful names
* Featurizer classes will name their generated features with the `prefix`, meaning the summary will contain their variable names as well.

The only case where no meaningful names will be retained is when a numpy array is fed into a featurizer (cannot concatenate numpy and pandas) or directly into a model. Here's an example:

```python
In [1]: from pmdarima.pipeline import Pipeline
   ...: from pmdarima.preprocessing import BoxCoxEndogTransformer, FourierFeaturizer
   ...: from pmdarima.arima import AutoARIMA
   ...: from pmdarima.datasets import load_wineind
   ...: import numpy as np
   ...:
   ...: wineind = load_wineind()
   ...: train, test = wineind[:125], wineind[125:]
In [2]: pipe = Pipeline([
   ...:         ("boxcox", BoxCoxEndogTransformer()),
   ...:         ("fourier", FourierFeaturizer(m=12)),
   ...:         ("arima", AutoARIMA(seasonal=False, stepwise=True,
   ...:                             suppress_warnings=True, d=1, max_p=2, max_q=0,
   ...:                             start_q=0, start_p=1,
   ...:                             maxiter=3, error_action='ignore'))
   ...:     ])
In [3]: pipe.fit(train)
Out[3]:
Pipeline(steps=[('boxcox',
                 BoxCoxEndogTransformer(floor=1e-16, lmbda=None, lmbda2=0,
                                        neg_action='raise')),
                ('fourier', FourierFeaturizer(k=None, m=12, prefix=None)),
                ('arima',
                 AutoARIMA(D=None, alpha=0.05, callback=None, d=1, disp=0,
                           error_action='ignore', information_criterion='aic',
                           m=1, max_D=1, max_P=2, max_Q=2, max_d=2,
                           max_order=10, max_p=2, max_q=0, maxiter=3,
                           method='lbfgs', n_fits=10, n_jobs=1,
                           offset_test_args=None, out_of_sample_size=0,
                           random=False, random_state=None, sarimax_kwargs=None,
                           scoring='mse', scoring_args=None, seasonal=False,
                           seasonal_test='ocsb', seasonal_test_args=None,
                           solver='lbfgs', ...))])
In [4]: pipe.summary()
Out[4]:
<class 'statsmodels.iolib.summary.Summary'>
"""
                           Statespace Model Results
==============================================================================
Dep. Variable:                      y   No. Observations:                  125
Model:               SARIMAX(2, 1, 0)   Log Likelihood                  49.601
Date:                Tue, 19 Nov 2019   AIC                            -67.202
Time:                        08:02:39   BIC                            -22.078
Sample:                             0   HQIC                           -48.872
                                - 125
Covariance Type:                  opg
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
intercept      0.0091      0.018      0.502      0.615      -0.026       0.044
FOURIER_0     -0.2289      0.024     -9.733      0.000      -0.275      -0.183
FOURIER_1      0.0247      0.023      1.075      0.282      -0.020       0.070
FOURIER_2     -0.1101      0.015     -7.209      0.000      -0.140      -0.080
FOURIER_3      0.1123      0.017      6.477      0.000       0.078       0.146
FOURIER_4     -0.2197      0.019    -11.530      0.000      -0.257      -0.182
FOURIER_5      0.2164      0.022     10.058      0.000       0.174       0.259
FOURIER_6     -0.0758      0.031     -2.424      0.015      -0.137      -0.015
FOURIER_7      0.0760      0.039      1.937      0.053      -0.001       0.153
FOURIER_8     -0.0060      0.020     -0.303      0.762      -0.045       0.033
FOURIER_9      0.1306      0.022      5.821      0.000       0.087       0.175
FOURIER_10  1739.6486   3461.000      0.503      0.615   -5043.786    8523.084
FOURIER_11     0.0467      0.026      1.808      0.071      -0.004       0.097
ar.L1         -0.7449      0.081     -9.235      0.000      -0.903      -0.587
ar.L2         -0.5253      0.089     -5.935      0.000      -0.699      -0.352
sigma2         0.0262      0.003      7.868      0.000       0.020       0.033
===================================================================================
Ljung-Box (Q):                       67.43   Jarque-Bera (JB):                30.98
Prob(Q):                              0.00   Prob(JB):                         0.00
Heteroskedasticity (H):               3.94   Skew:                             0.22
Prob(H) (two-sided):                  0.00   Kurtosis:                         5.41
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""
```

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation change

# How Has This Been Tested?

Multiple new unit tests, and existing unit tests pass

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pmdarima/preprocessing/exog/fourier.py [hunk 1]
```diff
@@ -60,6 +60,35 @@ class FourierFeaturizer(BaseExogFeaturizer, UpdatableMixin):
...
+    --------
+    >>> import pandas as pd
+    >>> pd.options.display.max_columns = 6  # for display...
+    >>> from pmdarima.preprocessing import FourierFeaturizer
+    >>> from pmdarima.datasets import load_wineind
+    >>> y = load_wineind()
+    >>> trans = FourierFeaturizer(12, 4)
+    >>> y_prime, exog = trans.fit_transform(y)
+    >>> exog.head()
+       FOURIER_0     FOURIER_1    ...      \
```
Comment (aaronreidsmith) [inline pmdarima/preprocessing/exog/fourier.py]: Does the output have that slash, or was that a product of copy/paste?

Comment (tgsmith61591) [reply pmdarima/preprocessing/exog/fourier.py]: It has that slash

Comment (aaronreidsmith) [reply pmdarima/preprocessing/exog/fourier.py]: Huh, looks funky to me, but as long as it's legit

**Comment by ihadanny** [comment]
nice! why aren't there 24 Fourier terms, sin and cos, with appropriate names?

**Comment by tgsmith61591** [comment]
> nice! why aren't there 24 Fourier terms, sin and cos, with appropriate names?

@ihadanny Because `("fourier", FourierFeaturizer(m=12))`. The period was 12. #227 added the "appropriate names".

**Comment by tgsmith61591** [comment]
From the doc:

```
    k : int, optional (default=None)
        The number of sine and cosine terms (each) to include. I.e., if ``k``
        is 2, 4 new features will be generated. ``k`` must not exceed ``m/2``,
        which is the default value if not set. The value of ``k`` can be
        selected by minimizing the AIC.
```

When `k` is None, it's set to `m // 2`:

```python
        m = self.m
        k = self.k
        if k is None:
            k = m // 2
```

So, with `m=12` and `k=None`, you'd expect 12 features (6 sin, 6 cos). We are busting our hides to fulfill these feature requests for free 🙂 Please recognize that.

**Comment by ihadanny** [comment]
thanks! I especially like the `d for double and also DUH` comment :)

**Comment by tgsmith61591** [comment]
Sometimes too much snark leaks into my code comments 😁

**Pull request merged**